### PR TITLE
WIP govc: Add virtual machine vApp configuration options

### PR DESCRIPTION
**Pull request by vancluever**
This PR (still WIP) adds/aims to add sub-commands to manage vApp settings on an individual virtual machine, useful for things like OVF authoring or if you want to use TF in its current state with OVAs that support things like cloud-init, ie: CoreOS - this allows you to remove vApp config so you can manage `guestinfo` directly. Of course, you could just use this to manage the VM directly too.

This also adds some flag helpers to the `flags` package, namely having to do with handling optional string values, string slices, and enumerations where validation is necessary (it doesn't look like vSphere properly kicks back an error on some improper vApp configuration options, like if you use lower case IPv4/IPv6 enumerations in the supported IP protocols). I didn't end up using some of these but I figured they'd be useful and will add tests to make sure they are properly accounted for.

Things present:

* Viewing of vApp configuration (`vm.vapp.info`)
* Editing of vApp configuration and product info (`vm.vapp.configure`)
* Viewing of vApp properties (`vm.vapp.property.info`)
* Addition of vApp properties (`vm.vapp.property.add`)
* Edit of vApp properties (`vm.vapp.property.edit`)
* Full removal of vApp configuration (`vm.vapp.remove`)

Things still remaining as of the evening of Jan 9th:

- [x]  Removal of vApp properties (`vm.vapp.property.remove`)
- [x] Addition of vApp handling to the simulator
- [x] Simulator tests
- [ ] Bats tests
- [ ] Addition of tests for the flag helpers

Related: #574

### CODE REVIEW

**Comment by vancluever** [comment]
I think I've gotten as far as I'm going to be able to get on this for the week (I have to turn my attention to TF stuff now).

This is functional, the govc commands work (tested them against our lab), the simulator code has been added but still needs to be tested, and the rest of the tests need to be in, which all in should hopefully not be a huge work commitment.

**Comment by dougm** [comment]
Awesome thanks @vancluever !  I will review by tomorrow.

File: govc/flags/enum.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+import (
+	"flag"
+	"fmt"
+	"strings"
+)
+
+// enum is a value store that validates against a set of strings. The
+// validation is performed at set-time.
+type enum struct {
```
Comment (dougm) [inline govc/flags/enum.go]: Nice.  We can make use of these in some of the other commands too.  At some point I'd like to ditch our custom CLI framework for something else, but not in a rush to do that.

File: govc/vm/vapp/configure.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+	spec := types.VirtualMachineConfigSpec{
+		VAppConfig: &cmd.VmConfigSpec,
+	}
+
+	task, err := vm.Reconfigure(ctx, spec)
+	if err != nil {
+		return err
+	}
+
+	if err := task.Wait(ctx); err != nil {
```
Comment (dougm) [inline govc/vm/vapp/configure.go]: Could use `cmd.ProgressLogger` here.

File: govc/vm/vapp/configure.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+type configure struct {
+	*flags.VirtualMachineFlag
+
+	types.VmConfigSpec
+
+	removeProduct bool
+}
+
+func init() {
+	cli.Register("vm.vapp.configure", &configure{})
```
Comment (dougm) [inline govc/vm/vapp/configure.go]: Other Reconfigure_Task based commands are named "change", `vm.change`, `cluster.change`, etc.  I wouldn't object to changing that name (no pun intended) in the future, but would prefer to keep it consistent for now.

File: govc/vm/vapp/property_add_edit.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+			Property: []types.VAppPropertySpec{op},
+		},
+	}
+
+	task, err := vm.Reconfigure(ctx, spec)
+	if err != nil {
+		return err
+	}
+
+	if err := task.Wait(ctx); err != nil {
```
Comment (dougm) [inline govc/vm/vapp/property_add_edit.go]: Could use cmd.ProgressLogger

File: simulator/virtual_machine_test.go [hunk 5]
```diff
@@ -758,3 +762,389 @@ func TestVmMarkAsTemplate(t *testing.T) {
...
+
+	vm.applyVAppPropertySpecRemove(2)
+	expected := "invalid property key for removal - removal key 2 not found\n"
+	actual := buf.String()
+	if expected != actual {
+		t.Fatalf("expected %q, got %q", expected, actual)
+	}
+}
+
+func testBoolPtr(b bool) *bool {
```
Comment (dougm) [inline simulator/virtual_machine_test.go]: Can use `types.NewBool`

**Comment by vancluever** [comment]
Thanks @dougm! Yeah I'm just trying to get coverage on the tests right now, balancing work/life at the same time. Hopefully will have more later this week.

Thanks for the note on the type helpers - I totally missed those - might be able to clean up some TF code with that too :slightly_smiling_face:

**Comment by vancluever** [comment]
Hey all, just wanted to mention I haven't forgotten about this, it just wasn't as urgent for our process as it was a while ago now that hashicorp/terraform#303 went live, and my personal time has been quite limited as of late. I still want to complete this though, and hopefully will have the spare time to do it soon!

**Comment by dougm** [comment]
@vancluever totally understand, take your time man.

**Comment by bpoland** [comment]
Hey @vancluever any chance you will be able to finish this off soon?  As you saw from https://github.com/terraform-providers/terraform-provider-vsphere/issues/489 I need to turn off vApp options on my templates (which get respun with security updates relatively frequently) and would love to be able to use govc to do that.  Thanks :)

**Comment by vancluever** [comment]
@bpoland I've been slowly working through my backlog of personal things, so hopefully I'll have a chance to finish this off soon! So yeah, I will be trying sometime in the next couple of weeks to close this off.

**Comment by bpoland** [comment]
Ok thanks @vancluever for this and all your work on the vsphere provider -- it is much appreciated!

**Comment by vancluever** [comment]
Hey everyone! I've started to work on this again - I'm just starting with addressing @dougm's review comments and finishing up the vApp property tests on the simulator, which should allow me to make progress on finishing up the rest of the simulator stuff soon. After that, I will move on to the rest of the testing stuff.

Thanks!

**Comment by vancluever** [comment]
Simulator tests are finally done! Moving on to bats tests now, in addition to looping back on an interface for adding EULAs. I added it to `vm.vapp.change`, but right now it's a command line argument with no easy way to view any EULAs that have been added. Since EULAs are generally large, this is kind of unwieldy right now.

**Comment by frapposelli** [comment]
@vancluever hi Chris, still interested in completing this PR?

**Comment by vancluever** [comment]
@frapposelli I think the truth of the matter is that I've had _very_ little personal time to devote to non-work related projects lately, so I probably won't have the bandwidth to do this anytime soon. I've actually been meaning to post a message about this for a while.

If anyone wants to take it over, I'm totally fine with that. Honestly the last things left to do to fix/add are the two remaining tasks described by the original comment (tests), and then it should be good to go.

**Comment by mylesagray** [comment]
Would still love to see this implemented - extremely useful for distros and cloud-images that take advantage of the `hostname` and `user-data` OVF Properties. Being able to change these on cloned VMs before boot would completely remove the need to rely on VMware Guest OS Customisation, which is a whole other layer of complexity.

**Comment by dougm** [comment]
@mylesagray is there a related CAPV issue open for this?

**Comment by petervansickel** [comment]
Any plans to get this completed?  I have a use case where I need to add vApp properties to a bunch of VMs as part of a Openshift cluster deployment on vSphere.

**Comment by piyushkv1** [comment]
When can we expect this to be completed?

**Comment by matteoiacovitti** [comment]
This will be amazing for my VM builds. I recently replaced packer by using the GOVC commands, except the one thing missing is modifying vApp options.

**Comment by mooperd** [comment]
+1

**Comment by embano1** [comment]
@dougm this PR is 3y old but there's still interest in the functionality. Since the PR itself is likely outdated, I suggest closing it and let interested users raise this in an issue for further discussion.

**Comment by vancluever** [comment]
Hey all, I'm closing this to clear out my old PR backlog. Will leave the repository open for now, but if anyone wants to carry on my work please clone it sooner rather than later as I may end up deleting my local fork.

**Comment by dougm** [comment]
@vancluever I just went to clone it now but your fork is gone.  Sorry we hadn't been able to pick up where you left off, but it still might be possible to if you can still make your changes available in some form if you still have it locally. Even patch in a gist would be good.  Thanks!

**Comment by vancluever** [comment]
@dougm I thought I still had this repo, but apparently I deleted it in a security cleanup over a month ago. Sorry 😞 I really meant to keep it.

I'm going to keep this bookmarked if I find it on an older workstation I'll make sure to re-push it.

Really sorry again!

**Comment by janitha** [comment]
The PRs content was still available in branch `pull/991/head` that you can explicitly fetch. I've placed it in my fork in branch `vapp` https://github.com/janitha/govmomi/tree/vapp for ease of access

**Comment by vancluever** [comment]
Thank you very much @janitha!!! Good to know this trick too just in case this happens again.

Thanks again!

**Comment by drennalls** [comment]
Is it possible to to remove all vapp properties from a VM without this PR?
OR (along the same lines) remove all properties from an OVA that's being imported?

**Pull request closed without merging**